File: leetcode/0102-binary-tree-level-order-traversal.cc
// https://leetcode.com/problems/binary-tree-level-order-traversal/
// Medium

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

#include<vector>

using namespace std;

class Solution {
    void levelOrderRec(TreeNode* root, vector<vector<int>>& ret, int level) {
        if (!root) return;
        if (ret.size() <= level) {
            ret.push_back(vector<int> {root->val});
        } else {
            ret[level].push_back(root->val);
        }
        levelOrderRec(root->left, ret, level + 1);
        levelOrderRec(root->right, ret, level + 1);
    }
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> v;
        levelOrderRec(root, v, 0);
        return v;
    }
};

File: leetcode/python/0102-binary-tree-level-order-traversal.py
# https://leetcode.com/problems/binary-tree-level-order-traversal/
# Medium

from collections import deque

# Better solution is to use BFS
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        ret = []
        q = deque([root])
        while len(q) > 0:
            l = len(q)
            ret.append([])

            for _ in range(l):
                node = q.popleft()
                ret[-1].append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        return ret

# Another solution is to use DFS where you keep track of the level
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ret = []
        
        def levelOrderRec(root: Optional[TreeNode], level: int):
            nonlocal ret
            if not root:
                return
            
            if level >= len(ret):
                ret.append([])
            ret[level].append(root.val)
            levelOrderRec(root.left, level+1)
            levelOrderRec(root.right, level+1)

        levelOrderRec(root, 0)
        return ret
